### DISCUSSION TITLE
ChartDataLabelsOptions/ChartDataLabelsDataset Does not work with complex data type

### DISCUSSION THREAD

**Discussion by Eltee-Taiwo** [Megabit/Blazorise#5144] [Q&A] [answered]
I have a bar chart that uses a complex type for it's underlying values and I want the chart data label to be a custom value as well. Ideally, something as simple as a properly formatted string but it doesn't look like that option exists.

```xml
<BarChart @ref="_barChartStacked" TItem="DataPoint" Options="_barChartOptions2">
    <ChartDataLabels TItem="DataPoint" Datasets="@_barDataLabelsDatasets" Options="@_barDataLabelsOptions" />
</BarChart>
```

This produces the following:
![image](https://github.com/Megabit/Blazorise/assets/37124491/ac35751d-2ed9-4db5-9b8d-e9bb56869fe3)

However, I want to be able to have a custom Label so I can have that text be whatever I want.

Here is my ChartDataLabelsDataset:

```csharp
readonly List<ChartDataLabelsDataset> _barDataLabelsDatasets = new()
{
    new()
    {
        DatasetIndex = 0,
        Options = new()
        {
            BackgroundColor = "#00ff00",
            BorderColor = "#00ff00",

        }
    },
    new ()
    {
        DatasetIndex = 1,
        Options = new ()
        {
            BackgroundColor = "Yellow",
            BorderColor = "Blue",
            // Labels = "Chicken" //I tried setting this property
        }
    }
};
```

I tried setting the Labels property on one of the datasets since the [chartjs plugin seems to check if that property exists](https://chartjs-plugin-datalabels.netlify.app/guide/formatting.html#data-transformation) but no dice.

Anyone have any thoughts on if this is possible? Or would this need to be a feature reuest?

**Comment by Eltee-Taiwo** [comment]
![image](https://github.com/Megabit/Blazorise/assets/37124491/de8b7f1a-6f0a-44ba-a9d4-2dd98a6701e0)

Playing around with the underlying chartjs plugin, I see that I can set up a custom formatter like so. So my question ultimately boils down to is this currently possible with this nuget package? Digging into it I see that the `Formatter` property only takes an enum that represents `Math.floor/ceeiling/round` functions.

**Reply by stsrki** [reply]
We already have an enum that is defined as 

```cs
Formater = ChartMathFormatter.Round,
```

Did you tried it?

---

In your screenshot, you are showing `formatter` as a JavaScript function. That is a little more tricky to do as JS functions must be executed on the client, not in C#. We have already done something similar in other Chart options where we transpile from C# to JS function. In case it is possible in this case we would have to introduce it in 1.4 or 2.0 as it could be somewhat of a breaking change.

**Reply by stsrki** [reply]
PS. maybe you can also try overriding `ToString()` on your complex object?

**Reply by Eltee-Taiwo** [reply]
Thanks for your input!

> Formater = ChartMathFormatter.Round,

Yes I tried that but that seems to only work with numbers - specifically decimals.

> Chart options where we transpile from C# to JS function

I was looking at how we currently do that [here](https://github.com/Megabit/Blazorise/blob/master/Source/Extensions/Blazorise.Charts.DataLabels/wwwroot/chart.datalabels.js#L55) and I want to see if I can maybe get that to work by changing the data type of [Formatter ](https://github.com/Megabit/Blazorise/blob/master/Source/Extensions/Blazorise.Charts.DataLabels/ChartDataLabelsOptions.cs#L56)in C# to use a scriptable option instead.

>  In case it is possible in this case we would have to introduce it in 1.4 or 2.0 as it could be somewhat of a breaking change.

Agreed. I think it will have to be a breaking change if this works. Unless we add another property called something like `ScriptableFormatter`

**Reply by Eltee-Taiwo** [reply]
https://github.com/Megabit/Blazorise/pull/5146

This is the best I could come up with at 4am but it works for my specific use case so I'll be moving this down my priotity list.

**Comment by Eltee-Taiwo** [comment, marked as answer]
https://github.com/Megabit/Blazorise/pull/5146

```csharp
readonly List<ChartDataLabelsDataset> _barDataLabelsDatasets = new()
{
    new()
    {
        DatasetIndex = 0,
        Options = new()
        {
            BackgroundColor = "#00ff00",
            BorderColor = "#00ff00",
            SoftCodeFormatter = "function(value, context) { return '£' + value.y; }"

        }
    },
    new ()
    {
        DatasetIndex = 1,
        Options = new ()
        {
            BackgroundColor = "Yellow",
            BorderColor = "Blue",
            SoftCodeFormatter = "function(value, context) { return '$' + value.y; }"
        }
    }
};
```

With the complex object looking like:
```csharp

    public class DataPoint
    {
        public int Y { get; set; }
        public string X { get; set; } = null!;
    }

```

Produces:
![image](https://github.com/Megabit/Blazorise/assets/37124491/2cae78f3-7cc1-4655-b4e7-76fa1ec10d69)
